### PR TITLE
Fusion: Add hint configuration

### PR DESCRIPTION
**Pull request by JeffGainsNGames**
adds hint configuration for Metroids and Charge Beam
image
image
image

### CODE REVIEW

**Comment by Miepee** [comment]
yakuza and restricted nav room text should be changed on the hints table to say something like "Charge Beam/Infant Metroids, Precision changeable"

File: randovania/games/fusion/exporter/hint_namer.py [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+    WHITE = "[COLOR=0]"
+    RED = "[COLOR=1]"
+    PINK = "[COLOR=2]"
+    YELLOW = "[COLOR=3]"
+    GREEN = "[COLOR=4]"
+    INDIGO = "[COLOR=5]"  # Poor readability
+    TEAL = "[COLOR=6]"
+
+
+def colorize_text(color: FusionColor, text: str, with_color: bool):
```
Comment (Miepee) [inline randovania/games/fusion/exporter/hint_namer.py:34]: mypy

File: randovania/games/fusion/exporter/hint_namer.py [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+        location_name = self.format_location(location, with_region=True, with_area=not hide_area, with_color=with_color)
+
+        return fmt.format(
+            colorize_text(self.color_item, resource.long_name, with_color),
+            determiner,
+            location_name,
+        )
+
+    def format_location_hint(self, game: RandovaniaGame, pick_hint: PickupHint, hint: Hint, with_color: bool) -> str:
+        msg = self.location_formatters[hint.precision.location].format(
```
Comment (Miepee) [inline randovania/games/fusion/exporter/hint_namer.py:103]: mypy

File: randovania/games/fusion/gui/preset_settings/fusion_hints_tab.py [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+
+    @classmethod
+    def tab_title(cls) -> str:
+        return "Hints"
+
+    @classmethod
+    def uses_patches_tab(cls) -> bool:
+        return False
+
+    def _on_art_combo_changed(self, new_index: int):
```
Comment (Miepee) [inline randovania/games/fusion/gui/preset_settings/fusion_hints_tab.py:37]: mypy

File: randovania/games/fusion/gui/preset_settings/fusion_hints_tab.py [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+        with self._editor as editor:
+            editor.set_configuration_field(
+                "hints",
+                dataclasses.replace(editor.configuration.hints, artifacts=self.hint_artifact_combo.currentData()),
+            )
+
+    def _on_cbeam_combo_changed(self, new_index: int):
```
Comment (Miepee) [inline randovania/games/fusion/gui/preset_settings/fusion_hints_tab.py:44]: mypy

File: randovania/games/fusion/gui/preset_settings/fusion_hints_tab.py [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+        with self._editor as editor:
+            editor.set_configuration_field(
+                "hints",
+                dataclasses.replace(editor.configuration.hints, charge_beam=self.charge_beam_hint_combo.currentData()),
+            )
+
+    def on_preset_changed(self, preset: Preset):
```
Comment (Miepee) [inline randovania/games/fusion/gui/preset_settings/fusion_hints_tab.py:51]: mypy

File: randovania/games/fusion/gui/preset_settings/fusion_hints_tab.py [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+        set_combo_with_value(self.hint_artifact_combo, preset.configuration.hints.artifacts)
```
Comment (Miepee) [inline randovania/games/fusion/gui/preset_settings/fusion_hints_tab.py:52]: mypy

File: randovania/games/fusion/gui/preset_settings/fusion_hints_tab.py [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+        set_combo_with_value(self.charge_beam_hint_combo, preset.configuration.hints.charge_beam)
```
Comment (Miepee) [inline randovania/games/fusion/gui/preset_settings/fusion_hints_tab.py:53]: mypy

File: randovania/games/fusion/exporter/hint_namer.py (hunk 2 above)
Comment (JeffGainsNGames) [reply randovania/games/fusion/exporter/hint_namer.py:103]: not sure how to reconcile this one

Comment (JeffGainsNGames) [reply randovania/games/fusion/exporter/hint_namer.py:103]: doesn't work, checked most other games and it appears they all have this same mypy error here.

Comment (Miepee) [reply randovania/games/fusion/exporter/hint_namer.py:103]: Oh, my bad lol
`hint.precision` is the one that has the `PrecisionPair | None` type. so you can:
```py
# either try this
prec = hint.precision
assert isinstance(prec, PrecisionPair)
# or try this
prec = typing.cast(PrecisionPair, hint.precision)
# end of block


msg = self.location_formatters[prec.location].format(
```

File: randovania/games/fusion/exporter/hint_namer.py [hunk 8]
```diff
@@ -0,0 +1,145 @@
+from __future__ import annotations
```
Comment (Miepee) [inline randovania/games/fusion/exporter/hint_namer.py:1]: it'd be nice to have tests for this file.

Comment (JeffGainsNGames) [reply randovania/games/fusion/exporter/hint_namer.py:1]: I unfortunately cannot find any comparable tests in other games.
What approximately needs to be done in the test and maybe I can take it from there

Comment (Miepee) [reply randovania/games/fusion/exporter/hint_namer.py:1]: just feed inputs for the important functions (format_resource_is_starting, format_guaranteed_resource, format_location_hint, format_region, format_area, format_player, format_joke) and check for a result

**Review by Miepee — APPROVED** [review]
lgtm

**Pull request merged**